File: stacks_and_queues/longest_subarray.java
public int longestSubarray(int[] nums, int limit)
{
    ArrayDeque<Integer> increasing = new ArrayDeque<>();
    ArrayDeque<Integer> decreasing = new ArrayDeque<>();

    int left = 0;
    int ans = 0;

    for (int right = 0; right < nums.length; right++)
    {
        // Maintain the monotonic deques
        while (!increasing.isEmpty() && increasing.getLast() > nums[right])
        {
            increasing.removeLast();
        }

        while (!decreasing.isEmpty() && decreasing.getLast() < nums[right])
        {
            decreasing.removeLast();
        }

        increasing.addLast(nums[right]);
        decreasing.addLast(nums[right]);

        // Maintain window property
        while (decreasing.getFirst() - increasing.getFirst() > limit)
        {
            if (nums[left] == decreasing.getFirst())
            {
                decreasing.removeFirst();
            }

            if (nums[left] == increasing.getFirst())
            {
                increasing.removeFirst();
            }

            left++;
        }

        ans = Math.max(ans, right - left + 1);
    }

    return ans;
}

File: stacks_and_queues/longest_subarray.js
var longestSubarray = function(nums, limit)
{
    let increasing = [];
    let decreasing = [];

    let left = 0, ans = 0;

    for (let right = 0; right < nums.length; right++)
    {
        // Maintain the monotonic deques
        while (increasing.length
               && increasing[increasing.length - 1] > nums[right])
        {
            increasing.pop();
        }

        while (decreasing.length
               && decreasing[decreasing.length - 1] < nums[right])
        {
            decreasing.pop();
        }

        increasing.push(nums[right]);
        decreasing.push(nums[right]);

        // Maintain window property
        while (decreasing[0] - increasing[0] > limit)
        {
            if (nums[left] == decreasing[0])
            {
                decreasing.shift();
            }

            if (nums[left] == increasing[0])
            {
                increasing.shift();
            }

            left++;
        }

        ans = Math.max(ans, right - left + 1);
    }

    return ans;
};